pkg/providers/gke.go

```diff
@@ -0,0 +1,342 @@
+		crToLB:        make(map[types.NamespacedName]types.NamespacedName),
+		lbToCRs:       make(map[types.NamespacedName]nameSet),
+		lbToPorts:     make(map[types.NamespacedName]int32Set),
+		capacityPerLB: capacity,
+	}
+	gke.client = initGCE()
+	gke.project = "kubecon-seattle-project"
+	gke.region = "us-central1"
+	return gke
+}
+
+func initGCE() *GCE {
+	c, err := google.DefaultClient(ctx, compute.CloudPlatformScope)
+	if err != nil {
+		log.WithName("gke").Error(err, "Failed to get GCE client")
+	}
+
+	computeService, err := compute.New(c)
+	if err != nil {
+		log.WithName("gke").Error(err, "Failed to create compute service")
+	}
+	
+	return &GCE{computeService}
+}
+
+func (i *GKE) GetCapacityPerLB() int {
+	return i.capacityPerLB
+}
+
+func (i *GKE) UpdateCache(key types.NamespacedName, lbSvc *corev1.Service) {
+	if lbSvc == nil {
+		delete(i.cacheMap, key)
+	} else {
+		if (len(lbSvc.Status.LoadBalancer.Ingress) > 0) {
+			//Make this IP Static and then add a Forwarding rule
+			fmt.Println("Updating Cache, let us make static, the ipaddress ",lbSvc.Status.LoadBalancer.Ingress[0].IP)
+			i.createStaticIP(lbSvc.ObjectMeta.Name+"-ip", lbSvc.Status.LoadBalancer.Ingress[0].IP)
+			//Let us also create a Forwarding rule to open up all ports.
+			//We can also do each port the slb needs but for now this is good enough.
+			fwRuleName := lbSvc.ObjectMeta.Name+"-rule"
+			lbPool := i.getTargetPool_usingLBName(lbSvc.ObjectMeta.Name)
+			if lbPool == nil {
+				log.WithName("GKE").Info("Cannot find the TargetPool for the LoadBalancer")
+			}
+			i.ForwardingRuleInsert(fwRuleName,lbSvc.Status.LoadBalancer.Ingress[0].IP, lbPool.SelfLink)
+		}
+		i.cacheMap[key] = lbSvc
+	}
+}
+
+func (i *GKE) NewService(sharedLB *kubeconv1alpha1.SharedLB) *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sharedLB.Name + SvcPostfix,
+			Namespace: sharedLB.Namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports:    sharedLB.Spec.Ports,
+			Selector: sharedLB.Spec.Selector,
+		},
+	}
+}
+
+func (i *GKE) NewLBService() *corev1.Service {	
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "lb-" + RandStringRunes(8),
+			Namespace: namespace,
+			Labels:    map[string]string{"lb-template": ""},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:     "tcp",
+					Protocol: corev1.ProtocolTCP,
+					Port:     33333,
+				},
+				// TODO(Huang-Wei): handle UDP case
+				// {
+				// 	Name:     "UDP",
+				// 	Protocol: corev1.ProtocolUDP,
+				// 	Port:     33333,
+				// },
+			},
+//			ExternalIPs: []string {
+//				ipaddress,
+//			},
+			Type: corev1.ServiceTypeLoadBalancer,
+		},
+	}
+}
+
+
+func (i *GKE) GetAvailabelLB(clusterSvc *corev1.Service) *corev1.Service {
+	// we leverage the randomness of golang "for range" when iterating
+OUTERLOOP:
+	for lbKey, lbSvc := range i.cacheMap {
+		if len(i.lbToCRs[lbKey]) >= i.capacityPerLB {
```

Comment:
This is a change I added for general purpose:

```suggestion
		if len(i.lbToCRs[lbKey]) >= i.capacityPerLB || len(lbSvc.Status.LoadBalancer.Ingress) == 0 {
```

Reply:
this is nice, thanks